### PR TITLE
Table of contents

### PR DESCRIPTION
**Pull request by wackerow**
## Description
- Fixes TableOfContentsLink logic when deciding which `id` to link to
- Removed redundant logic and isolated to `utils/toc.ts`
- Some file clean-up around ToC components and util functions
- Implemented ToC on Static layout

## Some preview URLs
- https://deploy-preview-44--ethereum-org-fork.netlify.app/about
- https://deploy-preview-44--ethereum-org-fork.netlify.app/glossary
- https://deploy-preview-44--ethereum-org-fork.netlify.app/history
- https://deploy-preview-44--ethereum-org-fork.netlify.app/defi
- https://deploy-preview-44--ethereum-org-fork.netlify.app/whitepaper
- https://deploy-preview-44--ethereum-org-fork.netlify.app/guides/how-to-swap-tokens

### CODE REVIEW

File: src/components/TableOfContents/index.tsx [hunk 1]
```diff
@@ -44,15 +44,11 @@ const TableOfContents: React.FC<IProps> = ({
 
   if (!isMobile) {
     const getTitleIds = (items: Array<ToCItem>, depth: number): void => {
```
Comment (nhsz) [inline src/components/TableOfContents/index.tsx:46]: should be moved to `src/lib/utils/toc.ts`

File: src/layouts/Static.tsx [hunk 2]
```diff
@@ -171,10 +173,14 @@ export const staticComponents = {
   YouTube,
 }
 
-export const StaticLayout = ({
+interface IProps extends PageContent, ChildOnlyProp {
```
Comment (nhsz) [inline src/layouts/Static.tsx]: IMO it'd be better to update the `PageContent` interface and add `lastUpdatedDate` prop as optional if its not required by all layouts, then remove `IProps`

File: src/components/UpgradeTableOfContents.tsx [hunk 3]
```diff
@@ -35,12 +32,6 @@ const TableOfContentsLink: React.FC<IPropsTableOfContentsLink> = ({ item: { titl
...
 }
 
-interface IPropsItemsList {
-  items: Array<ToCItem>
-  depth: number
-  maxDepth: number
-}
-
 const ItemsList: React.FC<IPropsItemsList> = ({ items, depth, maxDepth }) => {
   if (depth > maxDepth || !items) return null
```
Comment (nhsz) [inline src/components/UpgradeTableOfContents.tsx:37]: we could add a variable name explaining what this conditional means and use that inside the `if` instead, as its hard to follow what's happening

`x = depth > maxDepth || !items`, where `x` should be a declarative name explaining the condition

File: src/components/TableOfContents/index.tsx [hunk 4]
```diff
@@ -44,15 +44,11 @@ const TableOfContents: React.FC<IProps> = ({
...
-      if (depth > (maxDepth ? maxDepth : 1)) return
+      if (depth > (maxDepth ?? 1)) return
```
Comment (nhsz) [inline src/components/TableOfContents/index.tsx]: we could add a variable name explaining what this conditional means and use that inside the `if` instead, as its hard to follow what's happening

`x = depth > (maxDepth ?? 1)`, where `x` should be a declarative name explaining the condition

**Review by nhsz — CHANGES_REQUESTED** [review]
Good job @wackerow, asking for small adjustments and we're ready to merge

File: src/layouts/Static.tsx (hunk 2 above)
Comment (wackerow) [reply src/layouts/Static.tsx]: I'm game with that, can patch... Similarly I can update this on the other PRs.. I think it would affect the Frontmatter schema PR slightly

File: src/components/UpgradeTableOfContents.tsx (hunk 3 above)
Comment (wackerow) [reply src/components/UpgradeTableOfContents.tsx:37]: Okay.. I had trimmed the tertiary statement reflexively here, but careful of scope creep here (the logic here isn't new)... Similar to above, note that we had discussed not letting the relocation of all of these utils be a blocker for this migration.

File: src/components/TableOfContents/index.tsx (hunk 1 above)
Comment (wackerow) [reply src/components/TableOfContents/index.tsx:46]: See note below... I'm also realizing that this function uses locally scoped variables (ie. maxDepth) and uses side effects to push to the `titleIds` array. Would suggest we address relocating/refactoring this to a separate PR.

File: src/layouts/Static.tsx (hunk 2 above)
Comment (wackerow) [reply src/layouts/Static.tsx]: Oh oops, I actually read this backwards... @nhsz I agree, and have actually already implemented that approach in #42 (Frontmatter schema)... I'll patch this to match here.

File: src/components/UpgradeTableOfContents.tsx (hunk 3 above)
Comment (nhsz) [reply src/components/UpgradeTableOfContents.tsx:37]: Not talking about a relocation but extracting this conditional to a declarative variable to make code easier to follow, it's not trivial to understand what's happening without context. It'd help to add some comments too

File: src/components/TableOfContents/index.tsx (hunk 4 above)
Comment (nhsz) [reply src/components/TableOfContents/index.tsx]: Same as above

File: src/components/TableOfContents/index.tsx (hunk 1 above)
Comment (nhsz) [reply src/components/TableOfContents/index.tsx:46]: Ok let's leave this here

File: src/components/TableOfContents/index.tsx (hunk 4 above)
Comment (wackerow) [reply src/components/TableOfContents/index.tsx]: removed fallback (redundant) and added a comment

**Review by nhsz — APPROVED** [review]
LGTM!
